### PR TITLE
add linux support

### PR DESCRIPTION
**Pull request by gengjiawen**
Mainly switch to electron-builder. Related binary: https://github.com/gengjiawen/desktop/releases.

Things need to do: 
* I disable minify for production build
* I don't quick know how to pack a Mac Store build, this may need extra work.
* If you want to use travis ci to upload binary, please config the token. (For more info, please check https://github.com/SimulatedGREG/electron-vue/issues/141)

If you want me to make some change, please let me know.

### CODE REVIEW

**Comment by probonopd** [comment]
Would be awesome to see an AppImage build from this PR.

**Comment by gengjiawen** [comment]
This pr include binary for AppImage and deb.

**Comment by probonopd** [comment]
Unfortunately the build currently fails...

**Comment by gengjiawen** [comment]
On Mac and Windows, it seems like some test taking too much time. On linux, it seems test failed.

File: app/src/lib/open-shell.ts [hunk 1]
```diff
@@ -20,6 +20,11 @@ export function openShell(fullPath: string, shell?: string) {
     return spawn('gnome-terminal', commandArgs, { shell: true })
   }
 
+  if (__LINUX__) {
```
Comment (friederbluemle) [inline app/src/lib/open-shell.ts]: This looks identical to the `if` block above, so it probably shouldn't be here ;)

Comment (gengjiawen) [inline app/src/lib/open-shell.ts]: Yes, I have removed it.

**Comment by probonopd** [comment]
Now that they merged https://github.com/desktop/desktop/pull/2271, what does this mean for this PR? How can we get AppImages given that #2271 was merged?

**Comment by gengjiawen** [comment]
This pr replace electron-packager with electron-builder. You can directly build linux binary with npm task.

**Comment by probonopd** [comment]
I know but if they have merged the other PR then it seems like they want to use electron-packager rather than electron-builder.

**Comment by shiftkey** [comment]
@gengjiawen please let me know when you'd like me to look at this.

I'd like to see if we can leverage the `electron-packager` and one of the existing installer projects ([`electron-installer-debian`](https://github.com/unindented/electron-installer-debian), [`electron-installer-redhat`](https://github.com/unindented/electron-installer-redhat) and [`electron-installer-flatpak`](https://github.com/endlessm/electron-installer-flatpak) are mentioned in the [docs](https://github.com/electron-userland/electron-packager#distributable-creators)) before we consider switching over to `electron-builder` - we've invested a lot here, and I'd hate to break existing things unnecessarily.

Also, please have a look at https://github.com/shiftkey/desktop/pull/1 which is a working build and test setup for Desktop Linux on Travis - I'd like to get those incorporated sooner rather than later, so I might put together a PR for that this week on top of this discussion.

**Comment by gengjiawen** [comment]
My main consideration switch to `electron-builder` is that:
* It is simple to config multi-platform build.
* Use appveyor and travis-ci to automatically upload binary to Github Release.
* Very convenient to include static files.

I am not that familar with `electron-packager`, if we can do the linux build with `electron-packager` without much change, I am open to it :)

**Comment by Plasmmer** [comment]
Can you also create a snap?

**Comment by gengjiawen** [comment]
@shiftkey I am not familiar with the source code, for now I still can't let `npm run test:unit` get passed. Any idea why ?

**Comment by gengjiawen** [comment]
@Plasmmer I already made three linux binary type, do you have a strong reason why snap need supported ?

**Comment by Plasmmer** [comment]
Snap is the future of Ubuntu packages.

**Comment by shiftkey** [comment]
>  I still can't let `npm run test:unit` get passed. Any idea why ?

@gengjiawen it looks like `npm run test:unit` completes, but the next step `npm run test:integration` doesn't complete, and I think this is because of the changes to packaging.

For example, the launch test [is trying to use the contents of the `out` directory](https://github.com/desktop/desktop/blob/master/app/test/integration/launch-test.ts#L33) which probably doesn't exist now. It's probably the same for the other environments.

**Comment by gengjiawen** [comment]
But i fix the path in recent commit, maybe it's something else. I need to setup a linux environment in my local machine.

**Comment by shiftkey** [comment]
> But i fix the path in recent commit, maybe it's something else. 

Yes, you've updated the path. But if `electron-packager` is outputting installers rather than the loose files that were previously living in `out` then the integration tests are still going to fail.

**Comment by probonopd** [comment]
@gengjiawen it is possible to use `electron-builder` together with `electron-packager` to generate e.g., the AppImage: https://github.com/electron-userland/electron-builder#pack-only-in-a-distributable-format

**Comment by gengjiawen** [comment]
@shiftkey I know why the integration test failed, because it missing the git directory, I am planning add the copy process in `scrpit/test-setup.ts`.

**Comment by gengjiawen** [comment]
I find it, I missed the `package.json`.

**Comment by rogersachan** [comment]
Is there any status on this pull request?

**Comment by shiftkey** [comment]
@gengjiawen apologies for the delay in reviewing this. I'm worried about all of the changes underway here and how they affect the other platforms (currently Appveyor fails on linting, the latest Circle build fails on a unit test, and the Travis integration test fails). 

Now that we have first-class Travis support for Linux working in #2808, I've been throwing some spare cycles at how easily we can add packaging for whatever platforms we need. I know you are a fan of `electron-builder`, and I still don't like how much we have to depend on `electron-builder`, but you can see my progress over here: https://github.com/shiftkey/desktop/pull/2

Here's the package we can generate currently on Travis:



I'll outline why I want to continue down this path over this migration to `electron-builder`:

 - it doesn't currently require any changes to the contents under the `app` directory to support it. This might need to occur in the future as we uncover things that are platform-specific, and that's okay. But for now, we can keep things clean.
 - It keeps the build, test and package steps composable, rather than the unified approach of `electron-builder`
 - It integrates nicely with our how we package things for macOS and Windows, and gives us hooks to customize things if we need to.

I can see there's a number of merge conflicts in this PR due to the pace of changes with packages and environment changes, so I want an approach that makes it easier for forks like yours to keep up while we figure out what our official support looks like. **If I get these changes reviewed and merged, would they help with the heavy lifting of maintaining your fork of Desktop?**

**Comment by gengjiawen** [comment]
If you have any doubt, just leave this be. I will maintain this when I got time :)

**Comment by shiftkey** [comment]
Closing in favour of https://github.com/desktop/desktop/pull/2993

**Pull request closed without merging**